internal/receiver/smartagentreceiver/receiver.go

```diff
@@ -135,5 +148,79 @@ func (r *Receiver) createMonitor(monitorType string) (interface{}, error) {
 		output.AddExtraDimension(k, v)
 	}
 
-	return monitor, nil
+	// Configure SmartAgentConfigProvider to gather any global config overrides and
+	// set required envs.
+	configureEnvironmentOnce.Do(func() {
+		r.setUpSmartConfigProvider(extensions)
+		r.setUpEnvironment()
+	})
+
+	// Note, that this receiver has to configure main collectd even for collectd/custom monitor,
+	// despite the fact that, that monitor stands up its own instance of collectd to prevent this
+	// panic "Main collectd instance should not be accessed before being configured".
+	if r.config.monitorConfig.MonitorConfigCore().IsCollectdBased() {
+		configureCollectdOnce.Do(func() {
+			r.logger.Info("Configuring collectd")
+			err = configureMainCollectd(r.getCollectdConfig())
+		})
+	}
+
+	return monitor, err
+}
+
+func configureMainCollectd(collectdConfig *config.CollectdConfig) error {
+	return collectd.ConfigureMainCollectd(collectdConfig)
+}
+
+func (r *Receiver) setUpSmartConfigProvider(extensions map[configmodels.Extension]component.ServiceExtension) {
+	// If smartagent extension is not configured, use the default config.
+	f := smartagentextension.NewFactory()
+	defaultCfg := f.CreateDefaultConfig().(smartagentextension.SmartAgentConfigProvider)
+	r.config.saConfigProvider = defaultCfg
+
+	// Do a lookup for any smartagent extensions to pick up common collectd options
+	// to be applied across instances of the receiver.
+	for c := range extensions {
+		if c.Type() != f.Type() {
+			continue
+		}
+
+		cfgProvider, ok := c.(smartagentextension.SmartAgentConfigProvider)
+		if !ok {
+			continue
+		}
+		r.config.saConfigProvider = cfgProvider
+
+		// If there are multiple extensions configured, pick the first one. Ideally,
+		// there would only be one extension.
+		break
+	}
+}
+
+// getCollectdConfig returns a *config.CollectdConfig for r.config.collectdConfig.
+func (r *Receiver) getCollectdConfig() *config.CollectdConfig {
```

Comment:
Does this need to be in the receiver?  Seems to be an extension responsibility (and potentially unnecessary?)

Reply:
You're right. This should be moved to the extension. I need to update the extension's default to set `BundleDir`, `HasGenericJMXMonitor` and `IntervalSeconds` appropriately. I'll open a subsequent PR for updating the extension, removing this method.

Reply:
Addressed in #113

Reply:
Wouldn't it be easier to update the defaults now and just call `configureMainCollectd(saConfigProvider.getCollectdConfig())` instead of an intermediary phase/commit?

Reply:
sry, we raced each other...